### ISSUE TITLE
Single Fetch's useRouteLoaderData return type is missing undefined

### ISSUE THREAD

**Issue by skycoop** [remix-run/remix#9797]
### Reproduction

https://stackblitz.com/edit/remix-run-remix-zy1fcc?file=app%2Froot.tsx

### System Info

```shell
System:
    OS: Linux 6.5 KDE neon 6.0 6.0
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
    Memory: 1.87 GB / 7.55 GB
    Container: Yes
    Shell: 5.9 - /home/linuxbrew/.linuxbrew/bin/zsh
  Binaries:
    Node: 20.15.0 - ~/.nodenv/versions/20.15.0/bin/node
    npm: 10.7.0 - ~/.nodenv/versions/20.15.0/bin/npm
    pnpm: 9.6.0 - ~/.nodenv/versions/20.15.0/bin/pnpm
  npmPackages:
    @remix-run/dev: ^2.10.3 => 2.10.3 
    @remix-run/node: ^2.10.3 => 2.10.3 
    @remix-run/react: ^2.10.3 => 2.10.3 
    @remix-run/serve: ^2.10.3 => 2.10.3 
    vite: ^5.3.5 => 5.3.5
```

### Used Package Manager

pnpm

### Expected Behavior

The types in `single-fetch.d.ts` would suggest that the result of `useRouteLoaderData` is never undefined and therefore safe to destructure/use without testing for `undefined` first
https://github.com/remix-run/remix/blob/ede2012949a784722d8989124dff5a34ae90c008/packages/remix-react/future/single-fetch.d.ts#L27-L29
__

### Actual Behavior

If you navigate to a non-existent route in the reproduction, you'll get an error when `data.test` is accessed because the result of `useRouteLoaderData` was undefined.

**Comment by skycoop** [comment]
I'm not sure if the fix is as easy as updating the type in `single-fetch.d.ts` so I filed this as an issue instead of a PR, but totally willing to make a PR if it's that simple 😁

**Comment by ziimakc** [comment]
Maybe related, but I navigate to existing route: https://github.com/remix-run/remix/issues/9799

**Comment by brophdawg11** [comment]
👍 This was brought up in #9793 and will be fixed by #9796 in the next release
